File: myExpenses/src/main/java/org/totschnig/myexpenses/viewmodel/data/Currency.java
package org.totschnig.myexpenses.viewmodel.data;

import android.database.Cursor;
import android.os.Build;
import android.support.annotation.NonNull;

import com.google.auto.value.AutoValue;

import org.totschnig.myexpenses.model.CurrencyEnum;

import java.io.Serializable;

import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_CODE;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_LABEL;

@AutoValue
public abstract class Currency implements Serializable {
  public abstract String code();
  //should not count for equals
  private String displayName;
  final String displayName() {
    return displayName;
  }

  public static Currency create(@NonNull String code) {
    Currency currency = new AutoValue_Currency(code);
    currency.displayName = findDisplayName(code);
    return currency;
  }

  public static Currency create(Cursor cursor) {
    final String code = cursor.getString(cursor.getColumnIndexOrThrow(KEY_CODE));
    Currency currency = new AutoValue_Currency(code);
    currency.displayName = cursor.getString(cursor.getColumnIndexOrThrow(KEY_LABEL));
    if (currency.displayName == null) {
      currency.displayName = findDisplayName(code);
    }
    return currency;
  }

  private static String findDisplayName(String code) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
      try {
        return java.util.Currency.getInstance(code).getDisplayName();
      } catch (IllegalArgumentException ignored) {}
    }
    try {
      return CurrencyEnum.valueOf(code).getDescription();
    } catch (IllegalArgumentException ignored) {}
    return code;
  }

  public int sortClass() {
    switch (code()) {
      case "XXX":
        return 3;
      case "XAU":
      case "XPD":
      case "XPT":
      case "XAG":
        return 2;
      default:
        return 1;
    }
  }

  @Override
  public String toString() {
    return displayName();
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/viewmodel/data/Currency.kt
package org.totschnig.myexpenses.viewmodel.data

import android.content.Context
import android.database.Cursor
import android.os.Parcelable
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize
import org.totschnig.myexpenses.model.CommodityType
import org.totschnig.myexpenses.model.CurrencyEnum
import org.totschnig.myexpenses.provider.KEY_CODE
import org.totschnig.myexpenses.provider.KEY_COMMODITY_TYPE
import org.totschnig.myexpenses.provider.KEY_LABEL
import org.totschnig.myexpenses.provider.KEY_SYMBOL
import org.totschnig.myexpenses.provider.KEY_USAGES
import org.totschnig.myexpenses.provider.getEnum
import org.totschnig.myexpenses.provider.getIntIfExistsOr0
import org.totschnig.myexpenses.provider.getString
import org.totschnig.myexpenses.provider.getStringOrNull
import org.totschnig.myexpenses.util.Utils
import java.util.Locale

@Parcelize
data class Currency(
    val code: String,
    val displayName: String,
    val symbol: String = "¤",
    val usages: Int = 0,
    val fractionDigits: Int = 2,
    val commodityType: CommodityType = CommodityType.FIAT
) : Parcelable {
    @IgnoredOnParcel
    val sortClass = when (code) {
        "XXX" -> 3
        "XAU", "XPD", "XPT", "XAG" -> 2
        else -> 1
    }

    override fun toString() = displayName

    override fun hashCode() = code.hashCode()

    override fun equals(other: Any?) = when {
        this === other -> true
        other !is Currency -> false
        code != other.code -> false
        else -> true
    }

    companion object {
        fun create(code: String, context: Context) = create(code, Utils.localeFromContext(context))

        fun create(code: String, locale: Locale) = Currency(code, findDisplayName(code, locale))

        fun create(cursor: Cursor, locale: Locale): Currency {
            val code = cursor.getString(KEY_CODE)
            val label = cursor.getStringOrNull(KEY_LABEL)
            val usages = cursor.getIntIfExistsOr0(KEY_USAGES)
            val commodityType = cursor.getEnum(KEY_COMMODITY_TYPE, CommodityType.FIAT)
            val symbol = cursor.getString(KEY_SYMBOL)
            return Currency(
                code = code,
                displayName = label ?: findDisplayName(code, locale),
                symbol = symbol,
                usages = usages,
                commodityType = commodityType
            )
        }

        private fun findDisplayName(code: String, locale: Locale) = try {
            java.util.Currency.getInstance(code).getDisplayName(locale)
        } catch (_: IllegalArgumentException) {
            try {
                CurrencyEnum.valueOf(code).description
            } catch (_: IllegalArgumentException) {
                code
            }
        }
    }
}